x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/EsPhysicalOperationProviders.java

```diff
@@ -161,15 +161,15 @@ public Function<org.elasticsearch.compute.lucene.ShardContext, Query> querySuppl
     public final PhysicalOperation sourcePhysicalOperation(EsQueryExec esQueryExec, LocalExecutionPlannerContext context) {
         final LuceneOperator.Factory luceneFactory;
 
-        List<FieldSort> sorts = esQueryExec.sorts();
-        List<SortBuilder<?>> fieldSorts = null;
+        List<Sort> sorts = esQueryExec.sorts();
+        List<SortBuilder<?>> sortBuilders = null;
```

Comment:
I would have kept EsQueryExec the way it was and deal with geo-specific sorting here instead. This class is dealing with the translation of physical plan objects to actual operators (elements that do the actual physical work).
The LocalExecutionPlanner initiates the planning and coordinates the different physical plan components transformation and the PhysicalOperationProviders implementation deal with the creation of the operators for each specific physical plan. Having the specific implementation leak into EsQueryExec feels wrong.

Reply:
It feels like it already leaked into EsQueryExec. The fact that it contained a `FieldSort` which was specific to sorting on raw fields only, means that it was already not general to other kinds of sorts. If we did not add the geo sort info to EsQueryExec, we would need to attach it to something somewhere, and I cannot see where else it could be attached.

Reply:
I was thinking that perhaps what you are looking for is a way to not change the serialization of EsQueryExec, by keeping the GeoDistanceSort evaluation lower down, as you said _"here"_ in the `EsPhysicalOperationProviders`. The problem is that we can only push this down if we have the combination `TopNExec->EvalExec->EsQueryExec` with all the right conditions. This tree structure is not possible to determine here, and the only code that looks at the plan tree, and figures these things out is the `LocalPhysicalPlanOptimizer`, which is why we put it there.